src/semantic-highlighting.ts

```diff
@@ -0,0 +1,159 @@
+      for (const token of line.tokens) {
+        symbols.push({
+          id: 0,
+          kind: token.kind,
+          ranges: [Range.create(line.line, token.character, line.line, token.character + token.length)],
+          parentKind: 'Unknown', // FIXME
+          storage: 'None', // FIXME
+        });
+      }
+    }
+
+    const doc = workspace.getDocument(params.textDocument.uri);
+    await workspace.nvim.call('lsp_cxx_hl#hl#notify_symbols', [doc.bufnr, symbols]);
+  }
+
+  // Converts a string of base64 encoded tokens into the corresponding array of SemanticHighlightingToken.
+  private decodeTokens(tokens: string): SemanticHighlightingToken[] {
+    const scopeMask = 0xffff;
+    const lenShift = 0x10;
+    const uint32Size = 4;
+    const buf = Buffer.from(tokens, 'base64');
+    const retTokens: SemanticHighlightingToken[] = [];
+    for (let i = 0, end = buf.length / uint32Size; i < end; i += 2) {
+      const start = buf.readUInt32BE(i * uint32Size);
+      const lenKind = buf.readUInt32BE((i + 1) * uint32Size);
+      const scopeIndex = lenKind & scopeMask;
+      const len = lenKind >>> lenShift;
+      const kind = this.scopeTable[scopeIndex][0];
+      retTokens.push({ character: start, scopeIndex: scopeIndex, length: len, kind: this.decodeKind(kind) });
+    }
+
+    return retTokens;
+  }
+
+  private decodeKind(kind: string) {
+    // https://github.com/llvm/llvm-project/blob/9adc7fc3cdf571bd70d5f8bda4e2e9c233c5fd63/clang-tools-extra/clangd/SemanticHighlighting.cpp#L477
+    switch (kind) {
+      case 'entity.name.function.cpp':
+        return 'Function';
+      case 'entity.name.function.method.cpp':
+        return 'Method';
+      case 'entity.name.function.method.static.cpp':
+        return 'StaticMethod';
+      case 'variable.other.cpp':
+        return 'Variable';
+      case 'variable.other.local.cpp':
+        return 'LocalVariable';
+      case 'variable.parameter.cpp':
+        return 'Parameter';
+      case 'variable.other.field.cpp':
+        return 'Field';
+      case 'variable.other.field.static.cpp':
+        return 'StaticField';
+      case 'entity.name.type.class.cpp':
+        return 'Class';
+      case 'entity.name.type.enum.cpp':
+        return 'Enum';
+      case 'variable.other.enummember.cpp':
+        return 'EnumConstant';
+      case 'entity.name.type.typedef.cpp':
+        return 'Typedef';
+      case 'entity.name.type.dependent.cpp':
+        return 'DependentType';
+      case 'entity.name.other.dependent.cpp':
+        return 'DependentName';
+      case 'entity.name.namespace.cpp':
+        return 'Namespace';
+      case 'entity.name.type.template.cpp':
+        return 'TemplateParameter';
+      case 'entity.name.type.concept.cpp':
+        return 'Concept';
+      case 'storage.type.primitive.cpp':
+        return 'Primitive';
+      case 'entity.name.function.preprocessor.cpp':
+        return 'Macro';
+      case 'meta.disabled':
+        return 'InactiveCode';
```

Comment:
You need to send inactive regions to `lsp_cxx_hl#notify_skipped` which accepts a list of [`Ranges`](https://microsoft.github.io/language-server-protocol/specification#range).